designs/Path_Resolver_DESIGN.md

```diff
@@ -0,0 +1,118 @@
+        
+     def path(self):
+     	 path = ...
+     	 return path
+```
+
+We will also have a runtime validator class to make sure that the path to the executable, is actually the executable we want to be building artifacts for.
+This will do actions like making sure that version of the runtime specified, actually matches with the version from the executable path. Each Workflow can have its own validator that would do this.
+
+```python
+
+class RuntimeValidator(object):
+	...
+	...
+	def validate_runtime(self, runtime_path):
+		valid = ... (True/False based on some computation)
+		return valid
+```
+
+The Base workflow will define a ```get_executable``` and a  ```get_validator``` method.
+
+```python
+class BaseWorkflow(object):
+		...
+		...
+		...
+	
+	def get_executable(self):
+		return PathResolver(language=x, runtime=x.y).path
+	
+	def get_validator(self):
+		return RuntimeValidator(runtime_path=/usr/bin/x.y, language=x)
+		
+	@sanitize(executable_path=self.get_executable(),validator=self.get_validator())
+	def run(self):
+```
+
+This way we have de-coupled validtion of the path and the actual finding of the path to the executable. Both of these methods can be over-riden in any workflow that subclasses BaseWorkflow.
+
+There will be default implementations of ```PathResolver``` and ```RuntimeValidator``` that can still be used by workflow authors if they dont want to specialize them. 
+
+A decorator on top of run, would actually execute validation on the resolved path to ensure that its safe to actually start the workflow.
+
+### Implementation
+
+Here is an Example of Python workflow that has ```get_executable``` and ```get_validator``` defined.
+
+```python
+class PythonPipWorkflow(object):
+		...
+		...
+		...
+	
+	def get_executable(self):
+		return PythonPathResolver(language=python, runtime=python3.6).path
+	
+	def get_validator(self):
+		return PythonRuntimeValidator(runtime_path=/usr/bin/python3.6, language=python)
+```
+
+Finding the executable path and validation of the path occurs before the Workflow's ```run``` method is invoked. This way failure is detected early before workflow actions are executed.
+
+There is a work in progress PR that partially follows this design doc, except it adds the the runtime_path in the per-language workflow. This will be changed to make it align with this design doc.
+
+[#35](https://github.com/awslabs/aws-lambda-builders/pull/35)
+
+### Tenets
+
+The general tenets for this would look like as follows.
+
+1. Every workflow needs to have a PathResolver and a Validator. The resolver can be a custom version or use the bundled opioniated resolver.
+2. The workflow needs to fail if the resolved executable path does not match the supplied runtime version. This fails quickly and stops incorrect artifacts from being built.
+3. The resolved executable path should be available to every action.
+
+### Open questions
+
+* Should we constrain the interface for the PathResolver with an abstract base class?
```

Comment:
What is the pros and cons about this? I think we need an defined interface, at the very least.

Reply:
Pros of this, is that any implementations will need to have appropriate methods defined, so that we dont run into funkiness at runtime. But yeah, I'm just proposing abc as a way to solidly lock down the interface.